Fix gocyclo issue from clusterbootstrap controller

**Bug description**

There are linter issue from CI pipeline that we have to fix it as a followup.
```
Error: controllers/clusterbootstrap_controller.go:178:1: cyclomatic complexity 18 of func `(*ClusterBootstrapReconciler).reconcileNormal` is high (> 15) (gocyclo)
func (r *ClusterBootstrapReconciler) reconcileNormal(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
^
make: *** [Makefile:521: go-lint] Error 1
```

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [x] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Expected behavior**

no linter issue from ci pipeline

**Steps to reproduce the bug**

**Version** (include the SHA if the version is not obvious)

**Environment where the bug was observed (cloud, OS, etc)**

**Relevant Debug Output (Logs, manifests, etc)**

--- a/addons/controllers/clusterbootstrap_controller.go
+++ b/addons/controllers/clusterbootstrap_controller.go
@@ -175,8 +175,6 @@ func (r *ClusterBootstrapReconciler) Reconcile(ctx context.Context, req ctrl.Req
 }
 
 // reconcileNormal reconciles the ClusterBootstrap object
-// TODO: Fix the gocyclo linter issue: https://github.com/vmware-tanzu/tanzu-framework/issues/1761
-//nolint: gocyclo
 func (r *ClusterBootstrapReconciler) reconcileNormal(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
 	// get or clone or patch from template
 	clusterBootstrap, err := r.createOrPatchClusterBootstrapFromTemplate(cluster, log)
@@ -210,19 +208,7 @@ func (r *ClusterBootstrapReconciler) reconcileNormal(cluster *clusterapiv1beta1.
 		return ctrl.Result{}, err
 	}
 
-	if err := r.reconcileSystemNamespace(remoteClient); err != nil {
-		return ctrl.Result{}, err
-	}
-
-	// Create the ServiceAccount on remote cluster, so it could be referenced in PackageInstall CR for kapp-controller
-	// reconciliation.
-	if _, err := r.createOrPatchAddonServiceAccountOnRemote(cluster, remoteClient); err != nil {
-		return ctrl.Result{}, err
-	}
-
-	// Create the ClusterRole on remote cluster, and bind it to the ServiceAccount created in above. kapp-controller
-	// reconciliation needs privileges.
-	if err := r.createOrPatchAddonRBACOnRemote(cluster, remoteClient); err != nil {
+	if err := r.prepareRemoteCluster(cluster, remoteClient); err != nil {
 		return ctrl.Result{}, err
 	}
 
@@ -231,13 +217,14 @@ func (r *ClusterBootstrapReconciler) reconcileNormal(cluster *clusterapiv1beta1.
 	var corePackages []*runtanzuv1alpha3.ClusterBootstrapPackage
 	corePackages = append(corePackages, clusterBootstrap.Spec.CPI, clusterBootstrap.Spec.CSI)
 	corePackages = append(corePackages, clusterBootstrap.Spec.CNIs...)
+
+	// The following filtering out of nil items is not necessary in production
+	// as we do not expect CNI, CPI, CSI to be nil and webhook should handle
+	// the validations against those fields. This nil filter is mainly to allow
+	// local envtest run when any above component is missing.
+	corePackages = removeCorePackagesNils(corePackages)
+
 	for _, corePackage := range corePackages {
-		// The following nil check is redundant, we do not expect CNI, CPI, CSI to be nil and webhook should handle
-		// the validations against those fields. Having this nil check is mainly to allow local envtest run when
-		// any above component is missing.
-		if corePackage == nil {
-			continue
-		}
 		// There are different ways to have all the resources created or patched on remote cluster. Current solution is
 		// to handle packages in sequence order. I.e., Create all resources for CNI first, and then CPI, CSI. It is also
 		// possible to create all resources in a different order or in parallel. We will consider to use goroutines to create
@@ -552,6 +539,32 @@ func (r *ClusterBootstrapReconciler) reconcileSystemNamespace(clusterClient clie
 		r.Log.Info("created namespace", "namespace", r.Config.SystemNamespace)
 	}
 	return nil
+}
+
+func (r *ClusterBootstrapReconciler) prepareRemoteCluster(cluster *clusterapiv1beta1.Cluster, clusterClient client.Client) error {
+	if err := r.reconcileSystemNamespace(clusterClient); err != nil {
+		return err
+	}
+
+	// Create the ServiceAccount on remote cluster, so it could be referenced in PackageInstall CR for kapp-controller
+	// reconciliation.
+	if _, err := r.createOrPatchAddonServiceAccountOnRemote(cluster, clusterClient); err != nil {
+		return err
+	}
+
+	// Create the ClusterRole on remote cluster, and bind it to the ServiceAccount created in above. kapp-controller
+	// reconciliation needs privileges.
+	return r.createOrPatchAddonRBACOnRemote(cluster, clusterClient)
+}
+
+func removeCorePackagesNils(pkgs []*runtanzuv1alpha3.ClusterBootstrapPackage) []*runtanzuv1alpha3.ClusterBootstrapPackage {
+	var filtered []*runtanzuv1alpha3.ClusterBootstrapPackage
+	for _, pkg := range pkgs {
+		if pkg != nil {
+			filtered = append(filtered, pkg)
+		}
+	}
+	return filtered
 }
 
 // createOrPatchAddonServiceAccountOnRemote creates or patches the addon ServiceAccount on remote cluster.

